sycl/source/detail/composite_device/composite_device.cpp

```diff
@@ -0,0 +1,42 @@
+//==---------- composite_device.cpp - SYCL Composite Device ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <sycl/ext/oneapi/experimental/composite_device.hpp>
+#include <sycl/platform.hpp>
+
+namespace sycl {
+inline namespace _V1 {
+namespace ext::oneapi::experimental {
+std::vector<device> get_composite_devices() {
+  std::vector<device> Composites;
+  auto Devs = sycl::device::get_devices();
+  for (const auto &D : Devs) {
+    if (D.has(sycl::aspect::ext_oneapi_is_component)) {
+      auto Composite = D.get_info<info::device::composite_device>();
+      // Filter out duplicates.
+      if (std::find(Composites.begin(), Composites.end(), Composite) ==
+          Composites.end())
```

Comment:
nit: we have an `std::hash<device>` specialization, so I'd imagine `std::set` could work here. Feel free to ignore.